Import config.js file as ESM

paescuj opened the pull request:
Closes #19555

paescuj committed:
Import config.js file as ESM

paescuj committed:
Fix typo

paescuj committed:
Add changeset

paescuj committed:
Fix error message

br41nslug committed:
Merge branch 'main' into config-js-esm

br41nslug committed:
Fix import file paths for windows

br41nslug committed:
Merge branch 'config-js-esm' of https://github.com/directus/directus into config-js-esm

br41nslug commented on api/src/env.ts:
```diff
@@ -380,37 +381,36 @@ export const getEnv = () => env;
 
 	const fileExt = path.extname(configPath).toLowerCase();
 
 	if (fileExt === '.js') {
```
```suggestion
	if (['.js', '.cjs', '.mjs'].includes(fileExt)) {
```

let's allow for the CJS fallback same as for extensions.

paescuj committed:
Merge branch 'main' into config-js-esm

paescuj committed:
Support CJS

We now consider it a bug fix instead of a breaking change since it no
longer worked anyway and we keep support for CJS

azrikahar commented on api/src/env.ts:
```diff
@@ -380,37 +382,37 @@ export const getEnv = () => env;
-	if (fileExt === '.js') {
-		const module = require(configPath);
-		const exported = module.default || module;
+	if (['.js', '.cjs', '.mjs'].includes(fileExt)) {
```
Very side tangent here, but should we use this extensions constant over here?

https://github.com/directus/directus/blob/0c56c46fdf8dfd0a40ae04b3ec6270c42edf1077/packages/constants/src/files.ts#L1

this is a rather random question, shouldn't really block this PR at all as it works as is 👍

paescuj replied:
Nice 🚀❤️

br41nslug committed:
using constant for file extensions

paescuj commented on api/src/env.ts:
```diff
@@ -4,16 +4,22 @@
 
 const require = createRequire(import.meta.url);
 
+// casting as string[] so we can use .includes() properly
+const configFileExtensions = JAVASCRIPT_FILE_EXTS as unknown as string[];
+
```
```suggestion
```

paescuj commented on api/src/env.ts:
```diff
@@ -380,37 +386,37 @@ export const getEnv = () => env;
-	if (fileExt === '.js') {
-		const module = require(configPath);
-		const exported = module.default || module;
+	if (configFileExtensions.includes(fileExt)) {
```
```suggestion
	if ((JAVASCRIPT_FILE_EXTS as readonly string[]).includes(fileExt)) {
```

br41nslug replied:
I had it inlined at first but that makes it a lot less readable and does not leave us the option to explain why we did this weird trick. If we're going inline i think i'd rather type cast the extension like 
```suggestion
	if (JAVASCRIPT_FILE_EXTS.includes(fileExt as any)) {
```

paescuj replied:
Why `any`? `as readonly string[]` satisfies TS so we don't loose type safety and it's pretty self-explanatory, IMO.

br41nslug replied:
> Why any?

Because its the TS workaround that sticks closest to the approach you'd take if it wasnt a `readonly array` namely just `ARRAY.includes(value)`. Just because it has an `any` in there doesnt mean it loses type safety in any meaningful way in this case.

I strongly disagree about it being "self-explanatory" but it also doesnt really matter how we cast it the effect is the same, pick one 😄

paescuj replied:
- IMO, `any` should always be the last resort, it's also considered bad practice by `typescript-eslint`, see [`no-explicit-any`](https://typescript-eslint.io/rules/no-explicit-any) - we've disabled this rule for now (with the hope that we can clean this up someday):
  https://github.com/directus/directus/blob/0c56c46fdf8dfd0a40ae04b3ec6270c42edf1077/.eslintrc.js#L50-L51
- You lose type safety because you wouldn't notice anymore if `fileExt` changes, e.g. from `string` to `object`
- `as readonly string[]` only works because it's compatible with the type of `JAVASCRIPT_FILE_EXTS` so we don't lose type safety with this. Try to change it to e.g. `readonly number[]`, TS would complain.
- `(JAVASCRIPT_FILE_EXTS as readonly string[]).includes(fileExt)` feels self-explanatory to me because it shows:
  - `JAVASCRIPT_FILE_EXTS` it's most likely a constant
  - that is compatible with `readonly string[]` so we see it's really a read only const with string compatible content
  - it can be compared to `fileExt` so fileExt is a string

paescuj reviewed (commented):
I'd do the type assertion for `JAVASCRIPT_FILE_EXTS` inline since it's only used there and use `readonly` instead of `unknown` to satisfy the assertion.

paescuj committed:
Inline `JAVASCRIPT_FILE_EXTS`

u12206050 commented on api/src/env.ts:
```diff
@@ -362,7 +365,7 @@ const typeMap: Record<string, string> = {
 	...defaults,
 	...process.env,
-	...processConfiguration(),
+	...(await processConfiguration()),
```
This one await will impact all other files that import this to also become asynchronous.
This already hinders us from bundling Directus into a single executable. Recommend to encase this in an exported function instead.
